hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientInvocationServiceSupport.java

```diff
@@ -153,7 +158,19 @@ private void registerInvocation(ClientInvocation clientInvocation) {
         }
     }
 
+    private void checkMaxAllowedReached() {
+        if (maxAllowedConcurrentInvocations == Integer.MAX_VALUE) {
+            return;
+        }
+        long currentConcurrentInvocations = callIdIncrementer.get() - tailCallId.get();
+        if (maxAllowedConcurrentInvocations <= currentConcurrentInvocations) {
+            throw new HazelcastOverloadException("Max allowed concurrent invocations is reached : "
+                    + maxAllowedConcurrentInvocations);
+        }
+    }
+
     private ClientInvocation deRegisterCallId(long callId) {
+        tailCallId.incrementAndGet();
```

Comment:
Yes, you might be right. I think I need to decrease only if there is invocation on the map. But I want to think about this a little bit more.